[content-visibility] REGRESSION(267547@main): Blank panels on bing.com (content-visibility: auto)

rwlbuis opened the pull request:
#### 4241ce2896306ad1fc0443357a0ae93462fa7734
<pre>
[content-visibility] REGRESSION(267547@main): Blank panels on bing.com (content-visibility: auto)
<a href="https://bugs.webkit.org/show_bug.cgi?id=264989">https://bugs.webkit.org/show_bug.cgi?id=264989</a>

Reviewed by Simon Fraser.

When content-visibility status changes from hidden/out of view to visible/in viewport we should use dirtyVisibleContentStatus
since setHasVisibleContent has no effect on the subtree of the content-visibility root if the layer already has this flag set, causing
the subtree to remain hidden.

* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-096-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-096.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-097-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-097.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-098-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-098.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-099-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-099.html: Added.
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::styleWillChange):

Canonical link: <a href="https://commits.webkit.org/273399@main">https://commits.webkit.org/273399@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/79640d4761d1f291781d6eff2c52c3eab11abd89

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/35217 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/14150 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/37344 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/37972 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31784 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/36380 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/16531 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/11217 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/30680 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/35762 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11966 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/31405 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/10490 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/10546 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/31510 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/39220 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/32028 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/31841 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/36533 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10677 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/8607 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34545 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/12449 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/31164 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8080 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/11201 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/11510 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/RenderElement.cpp:
```diff
@@ -848,9 +848,12 @@ void RenderElement::styleWillChange(StyleDifference diff, const RenderStyle& new
-                else if (layer->hasVisibleContent() && (this == &layer->renderer() || layer->renderer().style().visibility() != Visibility::Visible))
+                if (willBeVisible) {
+                    if (m_style.visibility() != newStyle.visibility())
+                        layer->setHasVisibleContent();
```
Does this do the right thing if both visibility, and the skipped-ness changed?

rwlbuis replied:
I am not sure how to change the test (even locally) so that both happen in the same style update. But I changed the logic, since the dirtyVisibleContentStatus is more like a reset, testing the skipped content first should be safe, and even if the visibility becomes visible in addition to that, dirtyVisibleContentStatus should do the right thing (compared to setHasVisibleContent which has the early return case).

rwlbuis replied:
Ok, my modifications yesterday were wrong, it is possible to change visibility hidden to visible as well as unskipping skipped content in the same style update using the test below, I tried it and it passes (green rect on top is visible), since in this case setHasVisibleContent will not do an early return (m_hasVisibleContent is false).

```
<!doctype HTML>
<html class="reftest-wait">
<meta charset="utf8">
<title>Content Visibility: auto, scroll away and back</title>
<link rel="author" title="Rob Buis" href="mailto:rbuis@igalia.com">
<link rel="help" href="https://drafts.csswg.org/css-contain/#content-visibility">
<link rel="match" href="content-visibility-096-ref.html">
<meta name="assert" content="scolling away from c-v: auto container and back renders as expected">
<script src="/common/reftest-wait.js"></script>

<style>
.spacer {
  height: 10000px;
}
#child {
  width: 100px;
  height: 100px;
  background: green;
  position: relative
}
#target {
  width: 10px;
  height: 10px;
}
</style>

<div id="container" style="content-visibility: auto">
  <div id=child></div>
</div>
<div class=spacer></div>
<div id=target></div>

<script>

function step1() {
  document.getElementById("target").scrollIntoView(true /* alignToTop */);
  requestAnimationFrame(step2);
}

function step2() {
  container.style.visibility = "hidden";
  requestAnimationFrame(finishTest);
}

function finishTest() {
  container.setAttribute("style", "content-visibility: visible; visibility: visible");
  window.scrollTo(0, 0);
  requestAnimationFrame(takeScreenshot);
}

window.onload = requestAnimationFrame(() => {
  requestAnimationFrame(() => {
    requestAnimationFrame(() => {
      step1();
    });
  });
});

</script>
</html>
```

rwlbuis replied:
I now included above test as wpt test for completeness.

smfr commented on Source/WebCore/rendering/RenderElement.cpp:
```diff
@@ -853,9 +853,12 @@ void RenderElement::styleWillChange(StyleDifference diff, const RenderStyle& new
-                    layer->setHasVisibleContent();
-                else if (layer->hasVisibleContent() && (this == &layer->renderer() || layer->renderer().style().visibility() != Visibility::Visible))
+                if (willBeVisible) {
+                    if (m_style.skippedContentReason().has_value() && !newStyle.skippedContentReason().has_value())
```
If `willBeVisible` is true, we know `!newStyle.skippedContentReason().has_value()` is true here.

Do we need to think about change in the other direction (unskipped to skipped)?

rwlbuis replied:
Good point about not needing !newStyle.skippedContentReason().has_value() part of the check, removed.

I believe for unskipped to skipped, in the c-v: hidden case this is covered by the content-visibility-005.html test.

I think for c-v: auto it is less relevant since as soon as content is in the viewport it will count as unskipped, so I am not sure a "visual" test is possible. Let me know if you can think of a useful test scenario though.

rwlbuis replied:
Actually I did come up today with two tests for the unskipped to skipped scenario while rebasing the patch, PTAL :)

rwlbuis committed:
[content-visibility] REGRESSION(267547@main): Blank panels on bing.com (content-visibility: auto)
https://bugs.webkit.org/show_bug.cgi?id=264989

Reviewed by Simon Fraser.

When content-visibility status changes from hidden/out of view to visible/in viewport we should use dirtyVisibleContentStatus
since setHasVisibleContent has no effect on the subtree of the content-visibility root if the layer already has this flag set, causing
the subtree to remain hidden.

* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-096-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-096.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-097-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-097.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-098-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-098.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-099-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-099.html: Added.
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::styleWillChange):

Canonical link: https://commits.webkit.org/273399@main
